### PR TITLE
Add more tests for `apps/req`

### PR DESCRIPTION
**Pull request by romen**
https://github.com/openssl/openssl/issues/10224#issuecomment-546593113
highlighted that existing testing infrastructure is not covering common
usage patterns of the `req` app.

This commit explicitly adds request generations thorugh the CLI using
RSA, DSA and ECDSA (P-256) keys.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by romen** [comment]
WIP because merging this would break travis builds due to the bug mentioned in the description

- Ping @mattcaswell as he should know how to fix the bug before merging this
- Ping @levitte because my perl-foo sucks, so maybe there is a better way of adding these tests

File: test/recipes/25-test_req.t [hunk 1]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+            "-key", srctop_file("test", "testdsa.pem")])),
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
+           "Verifying signature on request");
+        }
+};
+
+subtest "generating certificate requests with DSA" => sub {
```
Comment (levitte) [inline test/recipes/25-test_req.t]: Shouldn't that title be different from the previous one?

File: test/recipes/25-test_req.t [hunk 2]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+subtest "generating certificate requests with RSA" => sub {
+    plan tests => 2;
+
+    SKIP: {
+        skip "RSA is not supported by this OpenSSL build", 2
+            if disabled("rsa");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testrsa.pem")])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: Nit: lines 57-58 should be aligned with the opening `[`

File: test/recipes/25-test_req.t [hunk 3]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: Same with line 62

File: test/recipes/25-test_req.t [hunk 4]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+subtest "generating certificate requests with DSA" => sub {
+    plan tests => 2;
+
+    SKIP: {
+        skip "DSA is not supported by this OpenSSL build", 2
+            if disabled("dsa");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testdsa.pem")])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: .. and lines 75-76

File: test/recipes/25-test_req.t [hunk 5]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: ... and line 80

File: test/recipes/25-test_req.t [hunk 6]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+subtest "generating certificate requests with ECDSA" => sub {
+    plan tests => 2;
+
+    SKIP: {
+        skip "ECDSA is not supported by this OpenSSL build", 2
+            if disabled("ec");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testec-p256.pem")])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: ... and line 93-94

File: test/recipes/25-test_req.t [hunk 7]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
```
Comment (levitte) [inline test/recipes/25-test_req.t]: ... and line 98

**Comment by levitte** [comment]
I'm actually fine with these tests looking the way they do, considering those that follow.  They can probably be improved to not be so massive copy-n-paste jobs, but that's a more major refactoring job that I think should be for another PR

File: test/recipes/25-test_req.t [hunk 8]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testrsa.pem")])),
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
+           "Verifying signature on request");
+        }
```
[the remark this answers is not in the document]
Comment (romen) [reply test/recipes/25-test_req.t]: Done

File: test/recipes/25-test_req.t [hunk 9]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testdsa.pem")])),
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
+           "Verifying signature on request");
+        }
```
[the remark this answers is not in the document]
Comment (romen) [reply test/recipes/25-test_req.t]: Done

File: test/recipes/25-test_req.t [hunk 10]
```diff
@@ -46,6 +46,60 @@ ok(!run(app([@addext_args, "-addext", $val, "-addext", $val2])));
...
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-new", "-out", "testreq.pem", "-utf8",
+            "-key", srctop_file("test", "testec-p256.pem")])),
+           "Generating request");
+
+        ok(run(app(["openssl", "req", "-config", srctop_file("test", "test.cnf"),
+            "-verify", "-in", "testreq.pem", "-noout"])),
+           "Verifying signature on request");
+        }
```
[the remark this answers is not in the document]
Comment (romen) [reply test/recipes/25-test_req.t]: Done

**Comment by romen** [comment]
@levitte indentation should now be fixed

the changes now involve also lines where only whitespace is changed as I noticed that the use of spaces vs tab for indentation was not uniform in the file

**Comment by levitte** [comment]
It looks like you found something, 'cause generating a req with DSA bombs:

```
/home/travis/build/openssl/openssl/_build/util/shlib_wrap.sh /home/travis/build/openssl/openssl/_build/apps/openssl req -config /home/travis/build/openssl/openssl/_srcdist/test/test.cnf -new -out testreq.pem -utf8 -key /home/travis/build/openssl/openssl/_srcdist/test/testdsa.pem => 139
```

Was this on purpose, i.e. this is a display of TDD (\*)?  Will there be a follow-up fixing PR?

-----

(\*) Test Driven Development

**Comment by levitte** [comment]
> the changes now involve also lines where only whitespace is changed as I noticed that the use of spaces vs tab for indentation was not uniform in the file

Very good catch.  The perl files certainly need some cleanup in that regard (you do *not* want to look at `Configure`)

**Comment by romen** [comment]
@levitte This was on purpose to catch automatically the second issue raised in #10224

The problem has been introduced whith the changes adding support for signatures through providers, that is why I suggested to fix the issue in #10303

**Comment by levitte** [comment]
Oh!  So it falls back on *me* :laughing:

**Comment by romen** [comment]
Or @mattcaswell :P

I don't know how to suggest a proper fix, because I don't know how in the new grand scheme of things we can ask the provider for the default message digest for the current key!

**Comment by levitte** [comment]
> ... I don't know how in the new grand scheme of things we can ask the provider for the default message digest for the current key!

As things currently stand, we can't, that's 100% legacy information for the moment, and of all places, it lies in the legacy `EVP_MD` definitions (code snippet from 1.1.1, this is the relevant part of `struct evp_md_st`):

https://github.com/openssl/openssl/blob/689c07b8d502301d82e09c1db04daf3c13b1d35d/include/crypto/evp.h#L116-L117

I'm not entirely sure how to tackle that in a sane way...

**Comment by levitte** [comment]
It looks like #10303 already fixes the problem...  so how about this PR is set to wait that we're done with #10303, gets rebased then and see what happens?

**Comment by romen** [comment]
Sounds good to me!

**Comment by romen** [comment]
Rebased on #10303 for testing

**Comment by romen** [comment]
It seems #10303 is fixing the tests added by this PR to highlight what described in https://github.com/openssl/openssl/issues/10224#issuecomment-546593113

There are `doc-nits` related red crossesdue to the changes in #10303 but otherwise the issue seems resolved once #10303 is merged.

**Comment by romen** [comment]
Rebased on top of merged #10303: it should now be ready for review @levitte

**Comment by levitte** [comment]
Keep an eye on the CIs, yeah?

**Comment by romen** [comment]
I think more testing never hurts, should this go also to 1.1.1?

(It cherry picks trivially, mostly because of the formatting)

**Comment by romen** [comment]
> Keep an eye on the CIs, yeah?

Yeah, I think the 24-hour rest between approval and ready-to-merge ensures also that we don't forget about that!

**Comment by petrovr** [comment]
Actually OpenSSL could generate ed25519 and ed448 certificates. Dunno who use them but maybe test could cover all possible cases.

**Comment by romen** [comment]
> Actually OpenSSL could generate ed25519 and ed448 certificates. Dunno who use them but maybe test could cover all possible cases.

@petrovr  Good point!
But I would prefer to add testing for EdDSA keys as a separate PR so that it's not limited to testing the `req` app, but extends to some of the other CLI utils.

I already started working on this, and while I can find normative test vectors for Ed25519 key material in https://tools.ietf.org/html/rfc8410#section-10 I can't find a good candidate for a Ed448 key for testing. Suggestions?

**Comment by romen** [comment]
@petrovr see #10410 for the PR adding more tests for EdDSA, thanks again for the remainder!

**Comment by romen** [comment]
Merged to `master` with b2a7310af0dd190712bae2e462a7708483dd4628 .

Thanks all!

**Pull request closed without merging**